### PR TITLE
Enable C structs AIP.

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: truffleruby/src/main/c/cext/ruby.c [hunk 1]
```diff
@@ -2291,23 +2291,34 @@ VALUE rb_get_path(VALUE object) {
...
 VALUE rb_struct_aset(VALUE s, VALUE idx, VALUE val) {
-  rb_tr_error("rb_struct_aset not implemented");
-  abort();
+  return (VALUE) truffle_invoke(RUBY_CEXT, "rb_struct_aset", s, idx, val);
 }
 
 VALUE rb_struct_define(const char *name, ...) {
-  rb_tr_error("rb_struct_define not implemented");
-  abort();
+  VALUE *rb_name = name == NULL ? truffle_read(RUBY_CEXT, "Qnil") : rb_str_new_cstr(name);
```
Comment (chrisseaton) [inline truffleruby/src/main/c/cext/ruby.c]: `truffle_read(RUBY_CEXT, "Qnil")` could just be `Qnil`. Don't worry about the extra function call overhead - we aren't doing performance yet, and anyway we have dynamic inlining!

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 2]
```diff
@@ -1152,6 +1152,25 @@ def rb_rescue2(b_proc, data1, r_proc, data2, rescued)
...
+    Struct.new(name, *attrs)
+  end
+
+  def rb_struct_aref(struct, index)
+    struct[index]
+  end
+
+  def rb_struct_aset(struct, index, value)
+    struct[index] = value
+  end
```
Comment (chrisseaton) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: These two could be done in C using `truffle_read` and `truffle_write`, which calls `[]` and `[]=` in Ruby automagically. But doesn't matter enough to change it now.

File: truffleruby/src/main/c/cext/ruby.c [hunk 3]
```diff
@@ -2291,23 +2291,34 @@ VALUE rb_get_path(VALUE object) {
...
+  VALUE *ary = rb_ary_new();
+  int i = 0;
+  char *arg = NULL;
+  while ((arg = (char *)truffle_get_arg(i + 1)) != NULL) {
```
Comment (eregon) [inline truffleruby/src/main/c/cext/ruby.c]: Is there a guarantee to get a final NULL argument at the end?

Comment (chrisseaton) [inline truffleruby/src/main/c/cext/ruby.c]: Yes - how else would Ruby know how many names to read?

Comment (eregon) [inline truffleruby/src/main/c/cext/ruby.c]: OK, just wanted to be sure it's not a hidden assumption on va_args or so.

**Comment by aardvark179** [comment]
Rebased and pushed locally since GitHub couldn't resolve the changes in cext.rb.

**Pull request closed without merging**